### PR TITLE
Initial PR for FairmotionInterface Class and Motion Viewer 

### PR DESCRIPTION
**Pull request by JuanTheEngineer**
## Motivation and Context

The goal of this PR is to push the first version of the FairmotionInterface Class and Motion Viewer with a few basic features to have a foundation for future feature PR's.

Here is a quick demo the amass male model performing actions from motion capture datasets. 

https://user-images.githubusercontent.com/59671006/137950099-57ef937a-b350-4051-a77c-a0312c0810d7.mp4

### How-to install
In your conda environment, you will need torch version 1.9.1
```
$ conda install pytorch=1.9.1 -c pytorch
``` 

Then, let's install Fairmotion and Human Body Prior from source.

To install fairmotion from source, first clone the git repository, use pip to download dependencies and build the project.
```
$ git clone https://github.com/facebookresearch/fairmotion.git
$ cd fairmotion
$ pip install -e .
```

Do the same with human body prior.
```
$ git clone https://github.com/nghorbani/human_body_prior.git
$ cd human_body_prior
$ pip install -e .
```

One last thing. At the moment of this merge, there was/is a bug in the fairmotion code. This [PR to fairmotion](https://github.com/facebookresearch/fairmotion/pull/55), when merged, will correct this bug. But if it is not merged yet, you will need to perform the one line change that can be found in the PR.

#### Motion data
For the motion data, we use npz mocap files from CMU datasets such as this one: `06_12_poses.npz` You can find these data sets by going to https://amass.is.tue.mpg.de/, making an account and logging in, and navigating the to Downloads > CMU > SMPL+H link that I have left [here](https://amass.is.tue.mpg.de/download.php#:~:text=551.56-,Body%20Data,-Renders)

#### Model data
Do the same to download the required body model. For SMPL-H download it from [here](http://mano.is.tue.mpg.de/). You will have to make a separate account for this site, although it looks similar to the Motion Data site. Once logged in, navigate to Downloads > Models & Code and click on "Extended SMPL+H model  (used in AMASS project)". Look for the file `smplh/male/model.npz`.
    
### Where to put the Data
By default, the Fairmotion Interface saves metadata files to `../habitat-sim/data/fairmotion/`, but you can set
a data directory by using the command-line argument `--metadata_dir` to change this directory. If the directory doesn't exist, the interface will simply create it and population it with a default json settings file. 
    
### How to run the viewer to see the motion
To help with running the Fairmotion Viewer Application, here is an example of what you could use.
```
# run with --help/-h for usage details
python examples/motion_viewer.py --dataset path/to/data/versioned_data/replica_cad_dataset_1.1/replicaCAD.scene_dataset_config.json --scene apt_0
```

Then, while the Fairmotion Viewer is running, you can press 'h' to get the list of key commands printed to the console. The `'f'` key will start the initially loaded motion.

### Features
- Loading mapped motion into human model
- Saving/fetching metadata
- Command-line arguments for loading motion files
- Key Frame and Trajectory Object previewing

## How Has This Been Tested

Building and running Fairmotion Viewer to test the FairmotionInterface API.

## Types of changes

- [x] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have completed my CLA (see **CONTRIBUTING**)
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by dhruvbatra** [comment]
This is awesome folks!

File: examples/fairmotion_interface.py [hunk 1]
```diff
@@ -0,0 +1,498 @@
...
+            # including FILE_SUFFIX if not included in name
+            filename = name + FILE_SUFFIX
+
+        filepath = f"{self.metadata_dir}{filename}.json"
+
+        try:
+            with open(filepath, "r") as file:
+                data = json.load(file)
+        except Exception:
+            raise Exception("Error: File does not appear to exist.")
```
[the remark this answers is not in the document]
Comment (aclegg3) [reply examples/fairmotion_interface.py:180]: We can keep it this way for now.

**Review by aclegg3 — APPROVED** [review]
Looks good. Excited for the follow-up features.

File: examples/fairmotion_interface.py [hunk 2]
```diff
@@ -0,0 +1,478 @@
...
+
+        # loading from file if given
+        # if file doesn't exist, make a new set of data with
+        # metadata_name as name
+        if metadata_name is not None:
+            try:
+                self.fetch_metadata(metadata_name)
+                self.set_data(name=metadata_name)
+            except Exception:
+                Exception(f"No file with name {metadata_name}, creating new file.")
```
Comment (Skylion007) [inline examples/fairmotion_interface.py:69]: @JuanTheEngineer @aclegg3 This is a bug. It does nothing. If you intended to raise it, the code after it it in the exception block wouldn't run. Did you mean to make this a logger.err?

File: examples/fairmotion_interface.py [hunk 3]
```diff
@@ -0,0 +1,478 @@
...
+        else:
+            # including FILE_SUFFIX if not included in name
+            filename = name + FILE_SUFFIX
+
+        filepath = f"{self.metadata_dir}{filename}.json"
+
+        try:
+            with open(filepath, "r") as file:
+                data = json.load(file)
+        except Exception:
```
Comment (Skylion007) [inline examples/fairmotion_interface.py:179]: These exception types are too broad? Handling a base exception like this is really bad. Did you intend to catch an `IOError`?

File: examples/fairmotion_interface.py [hunk 4]
```diff
@@ -0,0 +1,478 @@
...
+        metadata_name=None,
+        urdf_path=None,
+        amass_path=None,
+        bm_path=None,
+        metadata_dir=None,
+    ) -> None:
+        LoggingContext.reinitialize_from_env()
+        self.sim = sim
+        self.art_obj_mgr = self.sim.get_articulated_object_manager()
+        self.model: habitat_sim.physics.ManagedArticulatedObject = None
```
Comment (Skylion007) [inline examples/fairmotion_interface.py:48]:
```suggestion
        self.model: Optional[habitat_sim.physics.ManagedArticulatedObject] = None

File: examples/fairmotion_interface.py [hunk 5]
```diff
@@ -0,0 +1,478 @@
...
+        self.motion: motion.Motion = None
```
Comment (Skylion007) [inline examples/fairmotion_interface.py:49]:
```suggestion
        self.motion: Optional[motion.Motion] = None
```

**Comment by mmahdavian** [comment]
Hi. I receive following error after clicking on 'f' to load the model. Do you know what might be causing it?

```
[Sim] Fetched: data/fairmotion/default.json
[Sim] Loading...
[Sim] Done Loading.
[Sim] Loading...
[Sim] Done Loading.
[18:53:22:569676]:[Metadata] AttributesManagerBase.h(237)::loadAllFileBasedTemplates : Loading 1  Object Template templates found in ../habitat-sim/data/test_assets/objects
[18:53:22:569717]:[Metadata] AttributesManagerBase.h(352)::createFromJsonOrDefaultInternal : <Object Template>: Proposing JSON name : ../habitat-sim/data/test_assets/objects/sphere.object_config.json from original name : ../habitat-sim/data/test_assets/objects/sphere.object_config.json| This file exists.
[18:53:22:572001]:[Metadata] AbstractObjectAttributesManagerBase.h(211)::createObject : JSON Configuration File (../habitat-sim/data/test_assets/objects/sphere.object_config.json) based Object Template attributes created and registered.
[18:53:22:572025]:[Metadata] AttributesManagerBase.h(257)::loadAllFileBasedTemplates : <Object Template> : Loaded file-based templates: 1
[18:53:22:572108]:[Metadata] AttributesManagerBase.h(237)::loadAllFileBasedTemplates : Loading 1  Object Template templates found in ../habitat-sim/data/test_assets/objects
[18:53:22:572127]:[Metadata] AttributesManagerBase.h(352)::createFromJsonOrDefaultInternal : <Object Template>: Proposing JSON name : ../habitat-sim/data/test_assets/objects/nested_box.object_config.json from original name : ../habitat-sim/data/test_assets/objects/nested_box.object_config.json| This file exists.
[18:53:22:572191]:[Metadata] AbstractObjectAttributesManagerBase.h(211)::createObject : JSON Configuration File (../habitat-sim/data/test_assets/objects/nested_box.object_config.json) based Object Template attributes created and registered.
[18:53:22:572208]:[Metadata] AttributesManagerBase.h(257)::loadAllFileBasedTemplates : <Object Template> : Loaded file-based templates: 1
[18:53:22:587129]:[Nav] PathFinder.cpp(382)::build : Building navmesh with 145 x 260 cells
[18:53:22:627414]:[Nav] PathFinder.cpp(652)::build : Created navmesh with 125 vertices 63 polygons
[18:53:22:627455]:[Sim] Simulator.cpp(901)::recomputeNavMesh : reconstruct navmesh successful
[Sim] Command: load model
[Sim] Loading...
[18:53:27:186337]:[Physics] PhysicsManager.h(505)::addArticulatedObjectFromURDF : Not implemented in base PhysicsManager.
[18:53:27:186372]:[Core] ManagedContainerBase.h(204)::getObjectHandleByID : Unknown ArticulatedObject managed object ID: -1 . Aborting
[18:53:27:186395]:[Core] ManagedContainerBase.h(328)::checkExistsWithMessage : <ArticulatedObject>::getObjectCopyByID : Unknown ArticulatedObject managed object handle :  . Aborting
Traceback (most recent call last):
  File "examples/motion_viewer.py", line 779, in <module>
    FairmotionSimInteractiveViewer(sim_settings, fm_settings).exec()
  File "examples/motion_viewer.py", line 232, in key_press_event
    self.fm_demo.load_model()
  File "/home/mohammad/Mohammad_ws/habitat/habitat-sim/examples/fairmotion_interface.py", line 328, in load_model
    assert self.model.is_alive
AttributeError: 'NoneType' object has no attribute 'is_alive'
[18:53:27:327420]:[Physics] PhysicsManager.cpp(50)::~PhysicsManager : Deconstructing PhysicsManager
[18:53:27:328101]:[Scene] SceneManager.h(25)::~SceneManager : Deconstructing SceneManager
[18:53:27:328123]:[Scene] SceneGraph.h(25)::~SceneGraph : Deconstructing SceneGraph
GL::Context::current(): no current context
Aborted (core dumped)

```

**Comment by JuanTheEngineer** [comment]
@mmahdavian The first check I can think of is for you to make sure you have this file in this location `"../habitat-sim/data/test_assets/urdf/amass_male.urdf"`

@aclegg3 might be able to help you out further but my notes are inaccessible on the subject

**Comment by mmahdavian** [comment]
@JuanTheEngineer Thanks for responding. The amass_male.urdf is in the correct location. I don't know what is causing the error. @aclegg3 Do you have any idea about it?

**Comment by aclegg3** [comment]
> `PhysicsManager.h(505)::addArticulatedObjectFromURDF : Not implemented in base PhysicsManager.`
This tells me that physics is not enabled, so ArticulatedObjects are not available. Most likely cause is that Habitat-sim was not built with `--bullet`, so Bullet isn't installed and linked.

**Pull request merged**